File: python_expense_tracker/expense_operations.py
"""
Core business logic for Expense Tracking System
Implements all required core functions for expense management
"""

from typing import List, Dict, Optional, Any
from datetime import datetime
import models
import utils


def record_expense(date: str, category: str, title: str, amount: float, user_name: str) -> int:
    """
    Record a new expense with validation and user/category resolution
    
    Args:
        date: Expense date in YYYY-MM-DD format
        category: Category name
        title: Expense title/description
        amount: Expense amount
        user_name: Name of the user making the expense
        
    Returns:
        int: Expense ID of the created expense
        
    Raises:
        ValueError: If inputs are invalid
        sqlite3.IntegrityError: If user/category doesn't exist and can't be created
    """
    # Validate inputs
    date = utils.validate_date(date)
    amount = utils.validate_amount(str(amount))
    title = utils.validate_non_empty(title, "Title")
    user_name = utils.validate_non_empty(user_name, "User name")
    category = utils.validate_non_empty(category, "Category")
    
    # Get or create user
    user = models.get_user_by_name(user_name)
    if not user:
        try:
            user_id = models.create_user(user_name)
            print(f"Created new user: {user_name}")
        except Exception as e:
            raise ValueError(f"Failed to create user '{user_name}': {e}")
    else:
        user_id = user['id']
    
    # Get or create category
    category_obj = models.get_category_by_name(category)
    if not category_obj:
        try:
            category_id = models.create_category(category)
            print(f"Created new category: {category}")
        except Exception as e:
            raise ValueError(f"Failed to create category '{category}': {e}")
    else:
        category_id = category_obj['id']
    
    # Insert expense
    try:
        expense_id = models.insert_expense(date, category_id, title, amount, user_id)
        print(f"Expense recorded successfully with ID: {expense_id}")
        return expense_id
    except Exception as e:
        raise ValueError(f"Failed to record expense: {e}")


def view_expenses_by_date(min_date: Optional[str] = None, max_date: Optional[str] = None) -> List[Dict]:
    """
    View expenses filtered by date range
    
    Args:
        min_date: Minimum date (YYYY-MM-DD format)
        max_date: Maximum date (YYYY-MM-DD format)
        
    Returns:
        List[Dict]: List of expense dictionaries
    """
    try:
        # Validate dates if provided
        if min_date:
            min_date = utils.validate_date(min_date)
        if max_date:
            max_date = utils.validate_date(max_date)
        
        # Validate date range
        if min_date and max_date and min_date > max_date:
            raise ValueError("Start date cannot be after end date")
        
        expenses = models.fetch_expenses_by_filters(min_date=min_date, max_date=max_date)
        return expenses
    except Exception as e:
        print(f"Error filtering expenses by date: {e}")
        return []


def view_expenses_by_amount(min_amount: Optional[float] = None, max_amount: Optional[float] = None) -> List[Dict]:
    """
    View expenses filtered by amount range
    
    Args:
        min_amount: Minimum amount
        max_amount: Maximum amount
        
    Returns:
        List[Dict]: List of expense dictionaries
    """
    try:
        # Validate amounts if provided
        if min_amount is not None and min_amount < 0:
            raise ValueError("Minimum amount cannot be negative")
        if max_amount is not None and max_amount < 0:
            raise ValueError("Maximum amount cannot be negative")
        if min_amount is not None and max_amount is not None and min_amount > max_amount:
            raise ValueError("Minimum amount cannot be greater than maximum amount")
        
        expenses = models.fetch_expenses_by_filters(min_amount=min_amount, max_amount=max_amount)
        return expenses
    except Exception as e:
        print(f"Error filtering expenses by amount: {e}")
        return []


def view_expenses_by_category(categories: List[str]) -> List[Dict]:
    """
    View expenses filtered by category list
    
    Args:
        categories: List of category names to filter by
        
    Returns:
        List[Dict]: List of expense dictionaries
    """
    try:
        if not categories:
            return []
        
        # Get category IDs for the provided category names
        all_categories = models.get_all_categories()
        category_lookup = {cat['name'].lower(): cat['id'] for cat in all_categories}
        
        category_ids = []
        invalid_categories = []
        
        for category in categories:
            category_lower = category.lower()
            if category_lower in category_lookup:
                category_ids.append(category_lookup[category_lower])
            else:
                invalid_categories.append(category)
        
        if invalid_categories:
            print(f"Warning: Categories not found: {', '.join(invalid_categories)}")
        
        if not category_ids:
            print("No valid categories found.")
            return []
        
        expenses = models.fetch_expenses_by_filters(category_ids=category_ids)
        return expenses
    except Exception as e:
        print(f"Error filtering expenses by category: {e}")
        return []


def view_expenses_by_user(user_id: int) -> List[Dict]:
    """
    View expenses filtered by user ID
    
    Args:
        user_id: ID of the user to filter by
        
    Returns:
        List[Dict]: List of expense dictionaries
    """
    try:
        if user_id <= 0:
            raise ValueError("User ID must be positive")
        
        expenses = models.fetch_expenses_by_filters(user_id=user_id)
        return expenses
    except Exception as e:
        print(f"Error filtering expenses by user: {e}")
        return []


def view_all_expenses() -> List[Dict]:
    """
    View all expenses in the system
    
    Returns:
        List[Dict]: List of all expense dictionaries
    """
    try:
        expenses = models.fetch_expenses_by_filters()
        return expenses
    except Exception as e:
        print(f"Error retrieving all expenses: {e}")
        return []


def calculate_summary(expenses: Optional[List[Dict]] = None) -> Dict[str, Any]:
    """
    Calculate expense summary with totals and breakdowns
    
    Args:
        expenses: List of expenses to summarize (if None, uses all expenses)
        
    Returns:
        Dict: Summary dictionary with totals and breakdowns
    """
    try:
        if expenses is None:
            expenses = view_all_expenses()
        
        if not expenses:
            return {
                'total': 0.0,
                'count': 0,
                'by_category': {},
                'by_user': {}
            }
        
        # Calculate totals using list comprehensions (Python feature)
        total = sum(expense['amount'] for expense in expenses)
        count = len(expenses)
        
        # Calculate by category using dictionary comprehension (Python feature)
        by_category = {}
        for expense in expenses:
            category = expense['category_name']
            by_category[category] = by_category.get(category, 0) + expense['amount']
        
        # Calculate by user using dictionary comprehension (Python feature)
        by_user = {}
        for expense in expenses:
            user = expense['user_name']
            by_user[user] = by_user.get(user, 0) + expense['amount']
        
        return {
            'total': total,
            'count': count,
            'by_category': by_category,
            'by_user': by_user
        }
    except Exception as e:
        print(f"Error calculating summary: {e}")
        return {
            'total': 0.0,
            'count': 0,
            'by_category': {},
            'by_user': {}
        }


def get_expenses_by_date_range() -> List[Dict]:
    """
    Interactive function to get expenses by date range with user input
    
    Returns:
        List[Dict]: Filtered expenses
    """
    print("\n=== Filter Expenses by Date Range ===")
    min_date, max_date = utils.get_date_range_input()
    
    if min_date is None and max_date is None:
        print("No date range specified. Showing all expenses.")
        return view_all_expenses()
    
    return view_expenses_by_date(min_date, max_date)


def get_expenses_by_amount_range() -> List[Dict]:
    """
    Interactive function to get expenses by amount range with user input
    
    Returns:
        List[Dict]: Filtered expenses
    """
    print("\n=== Filter Expenses by Amount Range ===")
    min_amount, max_amount = utils.get_amount_range_input()
    
    if min_amount is None and max_amount is None:
        print("No amount range specified. Showing all expenses.")
        return view_all_expenses()
    
    return view_expenses_by_amount(min_amount, max_amount)


def get_expenses_by_category_interactive() -> List[Dict]:
    """
    Interactive function to get expenses by category with user input
    
    Returns:
        List[Dict]: Filtered expenses
    """
    print("\n=== Filter Expenses by Category ===")
    
    # Show available categories
    categories = models.get_all_categories()
    if not categories:
        print("No categories available.")
        return []
    
    print("Available categories:")
    for i, cat in enumerate(categories, 1):
        print(f"{i}. {cat['name']}")
    
    # Get user selection
    selected_categories = []
    while True:
        try:
            choice = input("\nEnter category number (or 'done' to finish): ").strip()
            if choice.lower() == 'done':
                break
            
            choice_num = int(choice)
            if 1 <= choice_num <= len(categories):
                selected_cat = categories[choice_num - 1]['name']
                if selected_cat not in selected_categories:
                    selected_categories.append(selected_cat)
                    print(f"Added category: {selected_cat}")
                else:
                    print(f"Category {selected_cat} already selected")
            else:
                print(f"Please enter a number between 1 and {len(categories)}")
        except ValueError:
            print("Please enter a valid number or 'done'")
        except KeyboardInterrupt:
            print("\nOperation cancelled.")
            return []
    
    if not selected_categories:
        print("No categories selected. Showing all expenses.")
        return view_all_expenses()
    
    return view_expenses_by_category(selected_categories)


def get_expenses_by_user_interactive() -> List[Dict]:
    """
    Interactive function to get expenses by user with user input
    
    Returns:
        List[Dict]: Filtered expenses
    """
    print("\n=== Filter Expenses by User ===")
    
    # Show available users
    users = models.get_all_users()
    if not users:
        print("No users available.")
        return []
    
    print("Available users:")
    for i, user in enumerate(users, 1):
        print(f"{i}. {user['name']}")
    
    # Get user selection
    try:
        choice = input(f"\nEnter user number (1-{len(users)}): ").strip()
        choice_num = int(choice)
        if 1 <= choice_num <= len(users):
            selected_user = users[choice_num - 1]
            return view_expenses_by_user(selected_user['id'])
        else:
            print(f"Please enter a number between 1 and {len(users)}")
            return []
    except ValueError:
        print("Please enter a valid number")
        return []
    except KeyboardInterrupt:
        print("\nOperation cancelled.")
        return []

File: cpp_expense_tracker/expense_operations.cpp
/**
 * expense_operations.cpp
 * Implementation of business logic
 * 
 * C++ FEATURES DEMONSTRATED:
 * - Exception handling (throw/catch vs Python's raise/except)
 * - STL algorithms and iterators
 * - Map operations for aggregation
 */

#include "expense_operations.h"
#include "models.h"
#include "utils.h"
#include <iostream>
#include <algorithm>

using namespace std;

/**
 * C++ FEATURE: Exception handling with try/catch
 * CONTRAST WITH PYTHON: Similar exception mechanism but different syntax
 * Python uses 'raise' and 'except', C++ uses 'throw' and 'catch'
 */
int ExpenseOperations::recordExpense(Database& db, const string& date,
                                    const string& category, const string& title,
                                    double amount, const string& user_name) {
    // Validate inputs
    if (!Utils::validateDate(date)) {
        throw invalid_argument("Invalid date format");
    }
    if (!Utils::validateAmount(amount)) {
        throw invalid_argument("Amount must be positive");
    }
    if (!Utils::validateNonEmpty(title)) {
        throw invalid_argument("Title cannot be empty");
    }
    if (!Utils::validateNonEmpty(user_name)) {
        throw invalid_argument("User name cannot be empty");
    }
    if (!Utils::validateNonEmpty(category)) {
        throw invalid_argument("Category cannot be empty");
    }
    
    // Get or create user
    int user_id = 0;
    auto user_opt = Models::getUserByName(db, user_name);
    
    if (user_opt.has_value()) {
        // C++ FEATURE: optional value extraction
        user_id = user_opt.value().id;
    } else {
        try {
            user_id = Models::createUser(db, user_name);
            cout << "Created new user: " << user_name << endl;
        } catch (const exception& e) {
            throw runtime_error("Failed to create user: " + string(e.what()));
        }
    }
    
    // Get or create category
    int category_id = 0;
    auto category_opt = Models::getCategoryByName(db, category);
    
    if (category_opt.has_value()) {
        category_id = category_opt.value().id;
    } else {
        try {
            category_id = Models::createCategory(db, category);
            cout << "Created new category: " << category << endl;
        } catch (const exception& e) {
            throw runtime_error("Failed to create category: " + string(e.what()));
        }
    }
    
    // Insert expense
    try {
        int expense_id = Models::insertExpense(db, date, category_id, title, amount, user_id);
        cout << "Expense recorded successfully with ID: " << expense_id << endl;
        return expense_id;
    } catch (const exception& e) {
        throw runtime_error("Failed to record expense: " + string(e.what()));
    }
}

vector<Expense> ExpenseOperations::viewExpensesByDate(Database& db,
                                                          const string& min_date,
                                                          const string& max_date) {
    try {
        // C++ FEATURE: Pointer to optional parameters
        const string* min_ptr = min_date.empty() ? nullptr : &min_date;
        const string* max_ptr = max_date.empty() ? nullptr : &max_date;
        
        return Models::fetchExpensesByFilters(db, min_ptr, max_ptr);
    } catch (const exception& e) {
        cerr << "Error filtering expenses by date: " << e.what() << endl;
        return vector<Expense>();
    }
}

vector<Expense> ExpenseOperations::viewExpensesByAmount(Database& db,
                                                            double min_amount,
                                                            double max_amount) {
    try {
        const double* min_ptr = (min_amount > 0.0) ? &min_amount : nullptr;
        const double* max_ptr = (max_amount > 0.0) ? &max_amount : nullptr;
        
        return Models::fetchExpensesByFilters(db, nullptr, nullptr, min_ptr, max_ptr);
    } catch (const exception& e) {
        cerr << "Error filtering expenses by amount: " << e.what() << endl;
        return vector<Expense>();
    }
}

vector<Expense> ExpenseOperations::viewExpensesByCategory(
    Database& db, const vector<string>& categories) {
    
    if (categories.empty()) {
        return vector<Expense>();
    }
    
    try {
        // Get all categories and build lookup map
        auto all_categories = Models::getAllCategories(db);
        
        // C++ FEATURE: STL map for lookups
        map<string, int> category_lookup;
        for (const auto& cat : all_categories) {
            category_lookup[Utils::toLower(cat.name)] = cat.id;
        }
        
        // Find category IDs
        vector<int> category_ids;
        for (const auto& cat_name : categories) {
            string lower_name = Utils::toLower(cat_name);
            auto it = category_lookup.find(lower_name);
            if (it != category_lookup.end()) {
                category_ids.push_back(it->second);
            }
        }
        
        if (category_ids.empty()) {
            cout << "No valid categories found." << endl;
            return vector<Expense>();
        }
        
        return Models::fetchExpensesByFilters(db, nullptr, nullptr, nullptr, nullptr, &category_ids);
    } catch (const exception& e) {
        cerr << "Error filtering expenses by category: " << e.what() << endl;
        return vector<Expense>();
    }
}

vector<Expense> ExpenseOperations::viewExpensesByUser(Database& db, int user_id) {
    if (user_id <= 0) {
        throw invalid_argument("User ID must be positive");
    }
    
    try {
        return Models::fetchExpensesByFilters(db, nullptr, nullptr, nullptr, nullptr, nullptr, &user_id);
    } catch (const exception& e) {
        cerr << "Error filtering expenses by user: " << e.what() << endl;
        return vector<Expense>();
    }
}

vector<Expense> ExpenseOperations::viewAllExpenses(Database& db) {
    try {
        return Models::fetchExpensesByFilters(db);
    } catch (const exception& e) {
        cerr << "Error retrieving all expenses: " << e.what() << endl;
        return vector<Expense>();
    }
}

/**
 * C++ FEATURE: STL map and algorithms for aggregation
 * CONTRAST WITH PYTHON: Python's dict.get() and list comprehensions are simpler
 * C++ requires more verbose iteration and map operations
 */
ExpenseSummary ExpenseOperations::calculateSummary(Database& db,
                                                  const vector<Expense>* expenses) {
    ExpenseSummary summary;
    
    try {
        vector<Expense> exp_list;
        
        if (expenses == nullptr) {
            exp_list = viewAllExpenses(db);
        } else {
            exp_list = *expenses;
        }
        
        if (exp_list.empty()) {
            return summary;
        }
        
        // Calculate totals
        summary.count = exp_list.size();
        
        // C++ FEATURE: STL algorithm for aggregation
        // CONTRAST: Python's sum() with generator expression is simpler
        summary.total = 0.0;
        for (const auto& expense : exp_list) {
            summary.total += expense.amount;
        }
        
        // Group by category
        for (const auto& expense : exp_list) {
            // C++ FEATURE: Map operator[] for auto-insertion
            summary.by_category[expense.category_name] += expense.amount;
        }
        
        // Group by user
        for (const auto& expense : exp_list) {
            summary.by_user[expense.user_name] += expense.amount;
            summary.user_expenses[expense.user_name].push_back(expense);
        }
        
        return summary;
    } catch (const exception& e) {
        cerr << "Error calculating summary: " << e.what() << endl;
        return summary;
    }
}

vector<Expense> ExpenseOperations::getExpensesByDateRange(Database& db) {
    cout << "\n=== Filter Expenses by Date Range ===" << endl;
    
    string min_date, max_date;
    if (!Utils::getDateRangeInput(min_date, max_date)) {
        return vector<Expense>();
    }
    
    if (min_date.empty() && max_date.empty()) {
        cout << "No date range specified. Showing all expenses." << endl;
        return viewAllExpenses(db);
    }
    
    return viewExpensesByDate(db, min_date, max_date);
}

vector<Expense> ExpenseOperations::getExpensesByAmountRange(Database& db) {
    cout << "\n=== Filter Expenses by Amount Range ===" << endl;
    
    double min_amount = 0.0, max_amount = 0.0;
    if (!Utils::getAmountRangeInput(min_amount, max_amount)) {
        return vector<Expense>();
    }
    
    if (min_amount == 0.0 && max_amount == 0.0) {
        cout << "No amount range specified. Showing all expenses." << endl;
        return viewAllExpenses(db);
    }
    
    return viewExpensesByAmount(db, min_amount, max_amount);
}

vector<Expense> ExpenseOperations::getExpensesByCategory(Database& db) {
    cout << "\n=== Filter Expenses by Category ===" << endl;
    
    // Show available categories
    auto categories = Models::getAllCategories(db);
    if (categories.empty()) {
        cout << "No categories available." << endl;
        return vector<Expense>();
    }
    
    cout << "Available categories:" << endl;
    for (size_t i = 0; i < categories.size(); ++i) {
        cout << (i + 1) << ". " << categories[i].name << endl;
    }
    
    // Get user selections
    vector<string> selected_categories;
    
    while (true) {
        string choice = Utils::getUserInput("\nEnter category number (or 'done' to finish): ");
        
        if (Utils::toLower(choice) == "done") {
            break;
        }
        
        try {
            int choice_num = stoi(choice);
            if (choice_num >= 1 && choice_num <= static_cast<int>(categories.size())) {
                string selected = categories[choice_num - 1].name;
                
                // Check if already selected
                if (find(selected_categories.begin(), selected_categories.end(), selected) 
                    == selected_categories.end()) {
                    selected_categories.push_back(selected);
                    cout << "Added category: " << selected << endl;
                } else {
                    cout << "Category " << selected << " already selected" << endl;
                }
            } else {
                cout << "Please enter a number between 1 and " << categories.size() << endl;
            }
        } catch (const exception&) {
            cout << "Please enter a valid number or 'done'" << endl;
        }
    }
    
    if (selected_categories.empty()) {
        cout << "No categories selected. Showing all expenses." << endl;
        return viewAllExpenses(db);
    }
    
    return viewExpensesByCategory(db, selected_categories);
}

vector<Expense> ExpenseOperations::getExpensesByUser(Database& db) {
    cout << "\n=== Filter Expenses by User ===" << endl;
    
    // Show available users
    auto users = Models::getAllUsers(db);
    if (users.empty()) {
        cout << "No users available." << endl;
        return vector<Expense>();
    }
    
    cout << "Available users:" << endl;
    for (size_t i = 0; i < users.size(); ++i) {
        cout << (i + 1) << ". " << users[i].name << endl;
    }
    
    // Get user selection
    try {
        string choice = Utils::getUserInput("\nEnter user number (1-" + 
                                                 to_string(users.size()) + "): ");
        int choice_num = stoi(choice);
        
        if (choice_num >= 1 && choice_num <= static_cast<int>(users.size())) {
            return viewExpensesByUser(db, users[choice_num - 1].id);
        } else {
            cout << "Please enter a number between 1 and " << users.size() << endl;
            return vector<Expense>();
        }
    } catch (const exception&) {
        cout << "Please enter a valid number" << endl;
        return vector<Expense>();
    }
}

